### DISCUSSION TITLE
Glasskube Native Packages should be written in ... ?

### DISCUSSION THREAD

**Discussion by pmig** [glasskube/glasskube#1018] [Polls]
Currently Glasskube only supports wrapping and patching Helm Charts and plain manifests. As [Helm chart creation has its quirks](https://glasskube.dev/blog/5-helm-shortcomings/#3-helm-chart-creation-is-not-user-friendly) creating native Glasskube packages is high up on our roadmap.

**Our main priorities are:**

-  Well defined package structure
-  Comprehensive but scaleable configuration options 
- Integration with existing typed configuration options & dependencies
- Support for in Code and optional dependencies
- Tooling that will package a Glasskube packages and generate manifests

_(this list will probably be changing over time)_

Please also let us know about the reasoning behind your answer:

**Comment by Jaykul** [comment]
Of those, I like Starlark and KCL the most. KCL is a more obvious fit in Kubernetes, and I think the only risk is confusion with all the methods KCL has for invoking their code in Kubernetes already. Starlark seems more flexible, and powerful, but I'm not sure if that's a good thing -- nothing against Tilt, but I'm not sure what I think of the idea of people writing whole libraries of Python functions for their deployment, so i think I'd prefer the simpler KCL option ;)

**Comment by linkdd** [comment]
I'm of the opinion that a configuration language should remain non turing complete, and that if you want what Starlark proposes, you're better off using Python anyway.

**NB:** As for writing whole libraries of Python for deployments, there is such a thing as [PyInfra](https://pyinfra.com/). I don't understand the popularity of Ansible when such tools exist :'( (and I want to say, Ansible is actually worse than Helm, because it turned YAML into an atrocious programming language with abysmal developer experience).

But beyond the "configuration" part, which only address "how to parametrize a package" (an important aspect, but not the only one), there is the actual deployment part.

What Helm does not communicate well, and I think it's one of the most important part of any deployment workflow, is the order in which Kubernetes resources are deployed. You want your CRDs to be deployed before your operator. You usually want your `ConfigMap` / `Secret` to be deployed before your `Deployment` or `Pods`, your `Service` before your `Ingress` (or the Gateway API equivalent), etc.

A declarative configuration language does not allow you to control this, you are relying on the choices of the underlying engine which uses that configuration. With a programming language however, you do have the control. What if I want to complete the deployment of this system (configmap + deployment + service + wait until ready) before deploying that other system, which are both part of the same packages ?

**NB:** This is at the moment impossible with Helm, you have to split your chart into smaller charts, which leaks implementation details about the "unit" you want to distribute. And then you have to use tooling like ArgoCD to orchestrate this. So the "package maintainer" delegates a whole lot of implementation details to the user.

---

**TL;DR:** Maybe a real programming language ?

**Comment by kosmoz** [comment]
Jsonnet could be another contender for the configuration language: https://jsonnet.org/
